src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2957 @@
+            var tupleType2 = model.GetTypeInfo(tuple2);
+            Assert.Null(tupleType2.Type);
+            Assert.Equal("(System.Object, ?)", tupleType2.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void TestFailedConversion()
+        {
+            var source = @"
+class C
+{
+    static void M(string s)
+    {
+        System.Console.Write((s, s) == (1, () => { }));
+    }
+}";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics(
+                // (6,30): error CS0019: Operator '==' cannot be applied to operands of type 'string' and 'int'
+                //         System.Console.Write((s, s) == (1, () => { }));
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "(s, s) == (1, () => { })").WithArguments("==", "string", "int").WithLocation(6, 30),
+                // (6,30): error CS0019: Operator '==' cannot be applied to operands of type 'string' and 'lambda expression'
+                //         System.Console.Write((s, s) == (1, () => { }));
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "(s, s) == (1, () => { })").WithArguments("==", "string", "lambda expression").WithLocation(6, 30)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var tuple1 = tree.GetCompilationUnitRoot().DescendantNodes().OfType<TupleExpressionSyntax>().ElementAt(0);
+            Assert.Equal("(s, s)", tuple1.ToString());
+            var tupleType1 = model.GetTypeInfo(tuple1);
+            Assert.Equal("(System.String, System.String)", tupleType1.Type.ToTestDisplayString());
+            Assert.Equal("(System.Object, System.String)", tupleType1.ConvertedType.ToTestDisplayString());
+
+            var tuple2 = tree.GetCompilationUnitRoot().DescendantNodes().OfType<TupleExpressionSyntax>().ElementAt(1);
+            Assert.Equal("(1, () => { })", tuple2.ToString());
+            var tupleType2 = model.GetTypeInfo(tuple2);
+            Assert.Null(tupleType2.Type);
+            Assert.Equal("(System.Object, ?)", tupleType2.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void TestDynamic()
+        {
+            var source = @"
+public class C
+{
+    public static void Main()
+    {
+        dynamic d1 = 1;
+        dynamic d2 = 2;
+        System.Console.Write($""{(d1, 2) == (1, d2)} "");
+        System.Console.Write($""{(d1, 2) != (1, d2)} "");
+
+        System.Console.Write($""{(d1, 20) == (10, d2)} "");
+        System.Console.Write($""{(d1, 20) != (10, d2)} "");
+    }
+}";
+            var comp = CreateStandardCompilation(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef, CSharpRef, SystemCoreRef }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "True False False True");
+        }
```

Comment:
Perhaps test `dynamic` and nested tuples: `dynamic d = (1, 2); bool b = (d, 3) == ((1, 2), 3);`

Reply:
Added `TestNestedDynamic`

---
In reply to: [170412986](https://github.com/dotnet/roslyn/pull/23356#discussion_r170412986) [](ancestors = 170412986)